### ISSUE TITLE
App loses communication with Extension

### ISSUE THREAD

**Issue by 2CNK** [limpkin/mooltipass#149]
I have a recurring issue that I discovered.  The MP seems to not wake up any more when the extension prompts to save new credentials.  Waking up manually (by finger presses) does not result in the MP saving new credentials from the extension, but disconnecting USB and re-connecting then unlocking card does.

I have recently uninstalled and rebooted and re-installed the application and extension to ensure no glitches.

Setup and problem:
- The MP has gone to sleep (screen is off).  
- MP app shows a cyan logo (connection ok?) and no error messages.
- When visiting a site and entering new credentials, MP extension correctly asks if I would like to add the credentials to the Mooltipass.  I agree.
- No response on the MP.  Touching the screen wakes it up but no prompt for saving or modifying any credentials.

Solution: 
- Unplug MP, plug back in.
- Enter PIN, get "Card unlocked" message.  MP app says "device status: unlocked".
- MP prompts to save credentials from earlier (the extension buffered it apparently)

Re-creation:
- Allow MP to go to sleep (screen off)
- Attempt to enter credentials somewhere so MP extension asks to save them

MP App version 0.10.0
MP Extention version 0.2.0
Windows 7, 64 bit

**Comment by limpkin** [comment]
Hello!

Who is it? What is your fw version?

Cheers

**Comment by 2CNK** [comment]
This is donp@aeinnovations.com - firmware version I flashed last is 0.7, but the extension says "unknown" for my firmware version when I look in the options for the extension.

**Comment by limpkin** [comment]
Is it also writen unknown in the app?

**Comment by 2CNK** [comment]
Ah.  No, it says "device is v0.7 32MBit" in the app.

The log shows no errors or anything, it just seems... delayed.  The extension buffers the request and the MP (once unplugged, plugged back in, and unlocked) just picks up where it left off.

Log sample (my comments in [])

device disconnected [unplugged from USB]
> looking for device
device found, connection made
device status: locked
device status: unlocked [pin entered OK]
> requesting to write credentials for smart-prototyping.com
> adding new service: smart-prototyping.com
> requesting to write credentials for smart-prototyping.com
> requesting to write credentials for smart-prototyping.com
credentials written
[time passes, MP goes to sleep.  New credential time with extension but nothing on MP even if woken up by finger presses.]
[unplug MP, wait a moment, plug in MP]
device disconnected
> looking for device
device found, connection made
device status: locked [enter PIN ok]
device status: unlocked
> requesting to write credentials for ebay.ca
> requesting to write credentials for ebay.ca
credentials written

And repeats just like this but with different site names.

**Comment by limpkin** [comment]
What do you mean by:
[time passes, MP goes to sleep. New credential time with extension but
nothing on MP even if woken up by finger presses.]

so the app doesn't get a "requesting to write credentials for..." ?

**Comment by 2CNK** [comment]
"New credential time" meaning I have visited another web site and entered new credentials on the site, the extension has detected this and asked if I want the MP to store the new credentials.

Here is the exact timing (I just did it)
1. Enter new credentials at site "boardgames.ca"
2. Extension asks do you want to save new credentials?
3. Yes (popup is dismissed)
4. Nothing on MP
5. Check app: no new log entries
6. Unplug MP
7. App has red icon.
8. App log says:
  device disconnected
  > looking for device
  device found, connection made
  > looking for device
9. Plug in MP
10. App has purple icon
11. App log appends:
 device found, connection made
 device status: locked
12. Enter PIN on MP
13. MP says "card unlocked"
14. App logo changes color to cyan
15. App log appends:
 device status: unlocked
 > requesting to write credentials for boardgames.ca
 > adding new service: boardgames.ca
16. MP screen prompts for adding new credentials.

**Comment by limpkin** [comment]
@kasbah: any reason why the app wouldn't see this second request?

**Comment by limpkin** [comment]
@kasbah can you add more debug info (for each request made to the mp) to see if it's a fw or app problem?

**Comment by uniacke1** [comment]
I may have experienced this too, I just chalked it up to the fact I've kept my computer in Standby and not done full shutdowns for ages...
Bjorn.

**Comment by uniacke1** [comment]
I may have experienced this too, I just chalked it up to the fact I've kept my computer in Standby and not done full shutdowns for ages...
Bjorn.

**Comment by limpkin** [comment]
not sure it is related but today i had more or less the same problem:

> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made
> looking for device
device found, connection made

**Comment by 2CNK** [comment]
Hmm, this is interesting.  No changes on my end since last posted.  However the following was noticed by me tonight.

The MP was frozen with screen on.  I unplugged and re-plugged the MP & unlocked but the App shows no response.  Closed and re-opened app, no response - stays at blank log and red icon.  Unplugged and re-plugged the MP again and unlocked, no response on App: red icon and blank log messages.

<pre>
1. Enter workshop and MP screen is on at "home" screen - it should be sleeping since has not been used.  
2. MP is not responsive to touches.
3. App (which is open) has cyan icon and no new log messages (no errors, etc.)
4. Unplug MP from USB, watch App to see if log messages or icon changes.
5. After a few moments of no change on the App, OS paints window with "Not responding"
6. Wait a bit longer.  App goes back to normal.  Icon remains cyan and unchanged.  No new log messages.
7. Plug MP back in.  No change on App.
8. Unlock MP.  Still no change on App.
9. Close and re-open App: red icon and blank log.
10. Unplug and re-plug in MP.  Unlock.  Still no change on App: red icon and blank log.
</pre>

**Comment by kasbah** [comment]
Thanks for the detailed reports. I just released the fix as 0.10.1 which you should be able to get through the Chrome Web Store within 60 mins.

**Comment by limpkin** [comment]
Your confidence is interesting....
Could you let us know what was happening ? :)

**Comment by kasbah** [comment]
Well, I was able to reproduce the issue reliably and then found a reliable fix. Not sure what's wrong with that. We can always re-open the issue if it hasn't been properly fixed :)

The problem was just that I forgot to take into account some "success states"  when determining if data needs to be sent to the device. The App used to always transitions out of these states  with each state update but with the last release I optimized the state update to happen only when inputs have changed so the App stayed in these "success states" and didn't realize it should send data. 

https://github.com/kasbah/mooltipass.hid-app/commit/f87a8ae0701c084981dc307158f896177371142d
